File: src/core/api/tauri/files.ts
import { invoke } from '@tauri-apps/api/core'
import type { DirectoryListing } from './types'

/**
 * Read a text file from the app data directory
 */
export const readTextFile = async (filePath: string): Promise<string> => {
  try {
    return await invoke<string>('read_text_file', { filePath })
  } catch (error) {
    console.error(`Failed to read file ${filePath}:`, error)
    throw error
  }
}

/**
 * Write a text file to the app data directory
 */
export const writeTextFile = async (filePath: string, content: string): Promise<void> => {
  try {
    await invoke<void>('write_text_file', { filePath, content })
  } catch (error) {
    console.error(`Failed to write file ${filePath}:`, error)
    throw error
  }
}

/**
 * List files in a directory within the app data directory
 */
export const listDirectory = async (dirPath: string): Promise<DirectoryListing> => {
  try {
    return await invoke<DirectoryListing>('list_directory', { dirPath })
  } catch (error) {
    console.error(`Failed to list directory ${dirPath}:`, error)
    throw error
  }
}

/**
 * Check if a file exists in the app data directory
 */
export const fileExists = async (filePath: string): Promise<boolean> => {
  try {
    return await invoke<boolean>('file_exists', { filePath })
  } catch (error) {
    console.error(`Failed to check if file exists ${filePath}:`, error)
    throw error
  }
}

/**
 * Utility function to ensure directory exists before writing
 */
export const ensureDirectoryExists = async (dirPath: string): Promise<void> => {
  try {
    // Try to list the directory to see if it exists
    await listDirectory(dirPath)
  } catch (error) {
    // If listing fails, the directory might not exist
    // We'll let the write operation create it
    console.warn(`Directory ${dirPath} might not exist, will be created on write`)
  }
}

/**
 * Read JSON file and parse it
 */
export const readJsonFile = async <T>(filePath: string): Promise<T> => {
  try {
    const content = await readTextFile(filePath)
    return JSON.parse(content) as T
  } catch (error) {
    console.error(`Failed to read JSON file ${filePath}:`, error)
    throw error
  }
}

/**
 * Write object as JSON file
 */
export const writeJsonFile = async <T>(filePath: string, data: T): Promise<void> => {
  try {
    const content = JSON.stringify(data, null, 2)
    await writeTextFile(filePath, content)
  } catch (error) {
    console.error(`Failed to write JSON file ${filePath}:`, error)
    throw error
  }
}

File: src-tauri/src/commands/files.rs
use crate::error::AppError;
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::Path;
use tauri::AppHandle;

#[derive(Debug, Serialize, Deserialize)]
pub struct FileInfo {
    pub name: String,
    pub path: String,
    pub size: u64,
    pub is_dir: bool,
    pub modified: Option<u64>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DirectoryListing {
    pub path: String,
    pub files: Vec<FileInfo>,
}

/// Read a text file from the app data directory
#[tauri::command]
pub async fn read_text_file(
    app_handle: AppHandle,
    file_path: String,
) -> Result<String, AppError> {
    // Resolve path relative to app data directory for security
    let app_data_dir = app_handle
        .path()
        .app_data_dir()
        .map_err(|e| AppError::FileError(format!("Failed to get app data directory: {}", e)))?;
    
    let full_path = app_data_dir.join(&file_path);
    
    // Ensure the path is within the app data directory (prevent path traversal)
    if !full_path.starts_with(&app_data_dir) {
        return Err(AppError::SecurityError("Path traversal attempt detected".to_string()));
    }
    
    fs::read_to_string(&full_path)
        .map_err(|e| AppError::FileError(format!("Failed to read file '{}': {}", file_path, e)))
}

/// Write a text file to the app data directory
#[tauri::command]
pub async fn write_text_file(
    app_handle: AppHandle,
    file_path: String,
    content: String,
) -> Result<(), AppError> {
    let app_data_dir = app_handle
        .path()
        .app_data_dir()
        .map_err(|e| AppError::FileError(format!("Failed to get app data directory: {}", e)))?;
    
    let full_path = app_data_dir.join(&file_path);
    
    // Ensure the path is within the app data directory
    if !full_path.starts_with(&app_data_dir) {
        return Err(AppError::SecurityError("Path traversal attempt detected".to_string()));
    }
    
    // Create parent directories if they don't exist
    if let Some(parent) = full_path.parent() {
        fs::create_dir_all(parent)
            .map_err(|e| AppError::FileError(format!("Failed to create directories: {}", e)))?;
    }
    
    fs::write(&full_path, content)
        .map_err(|e| AppError::FileError(format!("Failed to write file '{}': {}", file_path, e)))
}

/// List files in a directory within the app data directory
#[tauri::command]
pub async fn list_directory(
    app_handle: AppHandle,
    dir_path: String,
) -> Result<DirectoryListing, AppError> {
    let app_data_dir = app_handle
        .path()
        .app_data_dir()
        .map_err(|e| AppError::FileError(format!("Failed to get app data directory: {}", e)))?;
    
    let full_path = app_data_dir.join(&dir_path);
    
    // Ensure the path is within the app data directory
    if !full_path.starts_with(&app_data_dir) {
        return Err(AppError::SecurityError("Path traversal attempt detected".to_string()));
    }
    
    let entries = fs::read_dir(&full_path)
        .map_err(|e| AppError::FileError(format!("Failed to read directory '{}': {}", dir_path, e)))?;
    
    let mut files = Vec::new();
    
    for entry in entries {
        let entry = entry
            .map_err(|e| AppError::FileError(format!("Failed to read directory entry: {}", e)))?;
        
        let metadata = entry.metadata()
            .map_err(|e| AppError::FileError(format!("Failed to read file metadata: {}", e)))?;
        
        let file_name = entry.file_name().to_string_lossy().to_string();
        let relative_path = entry.path().strip_prefix(&app_data_dir)
            .unwrap_or(&entry.path())
            .to_string_lossy()
            .to_string();
        
        let modified = metadata.modified()
            .ok()
            .and_then(|time| time.duration_since(std::time::UNIX_EPOCH).ok())
            .map(|duration| duration.as_secs());
        
        files.push(FileInfo {
            name: file_name,
            path: relative_path,
            size: metadata.len(),
            is_dir: metadata.is_dir(),
            modified,
        });
    }
    
    Ok(DirectoryListing {
        path: dir_path,
        files,
    })
}

/// Check if a file exists in the app data directory
#[tauri::command]
pub async fn file_exists(
    app_handle: AppHandle,
    file_path: String,
) -> Result<bool, AppError> {
    let app_data_dir = app_handle
        .path()
        .app_data_dir()
        .map_err(|e| AppError::FileError(format!("Failed to get app data directory: {}", e)))?;
    
    let full_path = app_data_dir.join(&file_path);
    
    // Ensure the path is within the app data directory
    if !full_path.starts_with(&app_data_dir) {
        return Err(AppError::SecurityError("Path traversal attempt detected".to_string()));
    }
    
    Ok(full_path.exists())
}
